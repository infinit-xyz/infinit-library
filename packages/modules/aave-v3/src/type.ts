import { Address } from 'viem'

export type AaveV3Registry = {
  poolAddressProviderRegistry?: Address
  poolAddressesProvider?: Address
  liquidationLogic?: Address
  supplyLogic?: Address
  eModeLogic?: Address
  flashLoanLogic?: Address
  borrowLogic?: Address
  bridgeLogic?: Address
  poolLogic?: Address
  configuratorLogic?: Address
  reservesSetupHelper?: Address
  poolProxy?: Address
  lendingPools?: Record<string, LendingPool>
  walletBalanceProvider?: Address
  uiIncentiveDataProvider?: Address
  uiPoolDataProviderV3?: Address
  aaveProtocolDataProvider?: Address
  l2PoolImpl?: Address
  poolConfiguratorImpl?: Address
  aclManager?: Address
  aaveOracle?: Address
  aaveEcosystemReserveV2Proxy?: Address
  aaveEcosystemReserveController?: Address
  aaveEcosystemReserveV2Impl?: Address
  reserveInterestRateStrategies?: Record<string, Address>
  wrappedTokenGatewayV3?: Address
  l2Encoder?: Address
  aTokenImpl?: Address
  delegationAwareATokenImpl?: Address
  stableDebtTokenImpl?: Address
  variableDebtTokenImpl?: Address
  poolConfiguratorProxy?: Address
  rewardsControllerImpl?: Address
  rewardsControllerProxy?: Address
  pullRewardsTransferStrategy?: Address
  emissionManager?: Address
  aggregatorApi3Adapters?: Record<string, Address>
  aggregatorPythAdapters?: Record<string, Address>
  aggregatorBandAdapters?: Record<string, Address>
}

export type LendingPool = {
  underlyingToken: Address
  interestRateStrategy: Address
  aToken: Address
  stableDebtToken: Address
  variableDebtToken: Address
}

export type Modify<T, R> = Omit<T, keyof R> & R
