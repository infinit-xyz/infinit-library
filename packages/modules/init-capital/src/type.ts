import { Address } from 'viem'

export type InitCapitalRegistry = {
  proxyAdmin?: Address
  accessControlManager?: Address
  api3ProxyOracleReaderProxy?: Address
  initOracleProxy?: Address
  configProxy?: Address
  liqIncentiveCalculatorProxy?: Address
  posManagerProxy?: Address
  initCoreProxy?: Address
  riskManagerProxy?: Address
  lsdApi3ProxyOracleReaderProxy?: Address
  pythOracleReaderProxy?: Address
  moneyMarketHookProxy?: Address
  initLens?: Address
  api3ProxyOracleReaderImpl?: Address
  initOracleImpl?: Address
  configImpl?: Address
  liqIncentiveCalculatorImpl?: Address
  posManagerImpl?: Address
  initCoreImpl?: Address
  riskManagerImpl?: Address
  lsdApi3ProxyOracleReaderImpl?: Address
  pythOracleReaderImpl?: Address
  moneyMarketHookImpl?: Address
  lendingPoolImpl?: Address
  irms?: Record<string, Address>
  lendingPools?: Record<string, LendingPools>
}

export type LendingPools = {
  underlyingToken: Address
  lendingPool: Address
  irm: Address
}
