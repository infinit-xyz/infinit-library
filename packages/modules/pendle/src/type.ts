import { Address } from 'viem'

export type PendleV3Registry = {
  pendleRouterV4?: Address
  routerStorageV4?: Address
  actionAddRemoveLiqV3?: Address
  actionCallbackV3?: Address
  actionMiscV3?: Address
  actionSimple?: Address
  actionSwapPTV3?: Address
  actionSwapYTV3?: Address
  pendleRouterStatic?: Address
  actionStorageStatic?: Address
  actionInfoStatic?: Address
  actionMarketAuxStatic?: Address
  actionMarketCoreStatic?: Address
  actionMintRedeemStatic?: Address
  actionVePendleStatic?: Address
  pendleLimitRouter?: Address
  pendleLimitRouterProxy?: Address
  proxyAdmin?: Address
  baseSplitCodeFactoryContract?: Address
  oracleLib?: Address
  pendleGaugeControllerMainchainUpg?: Address
  pendleMarketFactoryV3?: Address
  pendlePYLpOracle?: Address
  pendleSwap?: Address
  pendleYieldContractFactory?: Address
  votingEscrowPendleMainchain?: Address
  pendleMsgSendEndpointUpgImpl?: Address
  pendleMsgSendEndpointUpgProxy?: Address
  pendleVotingContollerUpgImpl?: Address
  pendleVotingControllerUpgProxy?: Address
}
