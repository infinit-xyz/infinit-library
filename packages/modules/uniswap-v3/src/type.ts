import { Address } from 'viem'

export type UniswapV3Registry = {
  uniswapV3Factory?: Address
  nftDescriptor?: Address
  tickLens?: Address
  proxyAdmin?: Address
  swapRouter02?: Address
  quoterV2?: Address
  nonfungibleTokenPositionDescriptorImpl?: Address
  nonfungibleTokenPositionDescriptor?: Address
  nonfungiblePositionManager?: Address
  uniswapV3Staker?: Address
  permit2?: Address
  universalRouter?: Address
  unsupportedProtocol?: Address
}
