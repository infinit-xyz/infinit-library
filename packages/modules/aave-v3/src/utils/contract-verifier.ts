import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { AaveV3Registry, LendingPool } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class AaveV3ContractVerifier extends BaseContractVerifier<AaveV3Registry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: AaveV3Registry): Promise<ContractInfo[]> {
    const {
      lendingPools,
      reserveInterestRateStrategies,
      poolProxy,
      poolConfiguratorProxy,
      rewardsControllerProxy,
      api3Adapters,
      pythAdapters,
      bandAdapters,
      ...contractToAddressRegistry
    } = registry

    let contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (registry.poolAddressesProvider) {
      if (poolProxy) {
        contracts.push({
          address: poolProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
      if (poolConfiguratorProxy) {
        contracts.push({
          address: poolConfiguratorProxy,
          constructorArgs: [registry.poolAddressesProvider],
        })
      }
    }

    if (lendingPools && registry.poolConfiguratorProxy) {
      const lendingPoolContractInfos = await this.getLendingPoolContractInfos(registry.poolConfiguratorProxy, lendingPools)
      contracts = contracts.concat(lendingPoolContractInfos)
    }

    if (reserveInterestRateStrategies) {
      const reserveInterestRateStrategyContractInfos = await this.getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies)
      contracts = contracts.concat(reserveInterestRateStrategyContractInfos)
    }

    if (api3Adapters) {
      const api3AdaptersContractInfos = await this.getAdapterContractInfos(api3Adapters)
      contracts = contracts.concat(api3AdaptersContractInfos)
    }

    if (pythAdapters) {
      const pythAdaptersContractInfos = await this.getAdapterContractInfos(pythAdapters)
      contracts = contracts.concat(pythAdaptersContractInfos)
    }

    if (bandAdapters) {
      const bandAdaptersContractInfos = await this.getAdapterContractInfos(bandAdapters)
      contracts = contracts.concat(bandAdaptersContractInfos)
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }

  private async getLendingPoolContractInfos(poolConfiguratorProxy: Address, lendingPools: Record<string, LendingPool>) {
    const contracts: ContractInfo[] = []
    for (const info of Object.values(lendingPools)) {
      for (const [contractName, address] of Object.entries(info)) {
        if (contractName === 'underlyingToken') continue
        contracts.push({
          address,
          constructorArgs: contractName !== 'interestRateStrategy' ? [poolConfiguratorProxy] : undefined,
        })
      }
    }
    return contracts
  }

  private async getReserveInterestRateStrategyContractInfos(reserveInterestRateStrategies: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(reserveInterestRateStrategies)) {
      contracts.push({ address })
    }
    return contracts
  }

  private async getAdapterContractInfos(adapters: Record<string, Address>) {
    const contracts: ContractInfo[] = []
    for (const address of Object.values(adapters)) {
      contracts.push({ address })
    }
    return contracts
  }
}
