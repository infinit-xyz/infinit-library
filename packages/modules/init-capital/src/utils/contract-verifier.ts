import { Address, PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { InitCapitalRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<InitCapitalRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: InitCapitalRegistry): Promise<ContractInfo[]> {
    const { lendingPools, irms, ...contractToAddressRegistry } = registry
    const contracts: ContractInfo[] = []

    for (const address of Object.values(contractToAddressRegistry)) {
      contracts.push({ address })
    }

    if (lendingPools) {
      for (const { lendingPool, irm } of Object.values(lendingPools)) {
        contracts.push({ address: lendingPool })
        contracts.push({ address: irm })
      }
    }

    if (irms) {
      for (const address of Object.values(irms)) {
        contracts.push({ address })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
