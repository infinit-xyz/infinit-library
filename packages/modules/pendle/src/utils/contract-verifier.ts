import { PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { PendleRegistry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class TokenContractVerifier extends BaseContractVerifier<PendleRegistry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(_registry: PendleRegistry): Promise<ContractInfo[]> {
    const contracts: ContractInfo[] = []

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
