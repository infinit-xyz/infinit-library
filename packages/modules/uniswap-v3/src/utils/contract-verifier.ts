import { PublicClient } from 'viem'

import { BaseContractVerifier, BlockExplorerParams, ContractInfo } from '@infinit-xyz/core/internal'

import { Artifacts } from 'hardhat/types'

import { UniswapV3Registry } from '@/src/type'
import { getArtifacts } from '@/src/utils/artifact'

export class UniswapV3ContractVerifier extends BaseContractVerifier<UniswapV3Registry> {
  constructor(client: PublicClient, params: BlockExplorerParams) {
    super(client, params)
  }

  protected override async getContracts(registry: UniswapV3Registry): Promise<ContractInfo[]> {
    const { pools, ...rest } = registry

    const contracts: ContractInfo[] = []

    for (const address of Object.values(rest)) {
      contracts.push({ address })
    }

    if (pools) {
      for (const pool of pools) {
        contracts.push({ address: pool, constructorArgs: [] })
      }
    }

    return contracts
  }

  protected override async getArtifacts(): Promise<Artifacts> {
    const artifacts = await getArtifacts()

    return artifacts
  }
}
