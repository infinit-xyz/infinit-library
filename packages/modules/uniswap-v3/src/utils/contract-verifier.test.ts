import { describe, expect, test } from 'vitest'

import { createPublicClient, http } from 'viem'
import { holesky } from 'viem/chains'

import { UniswapV3Registry } from '../type'
import { UniswapV3ContractVerifier } from './contract-verifier'

// Create a subclass to expose the protected method for testing
class TestUniswapV3ContractVerifier extends UniswapV3ContractVerifier {
  public async testGetContracts(registry: UniswapV3Registry) {
    return this.getContracts(registry)
  }
}

describe('UniswapV3ContractVerifier.ts', () => {
  const registry: UniswapV3Registry = {
    uniswapV3Factory: '0x8f0c29795615148f30bd402876fa69cbf88a91d4',
    nftDescriptor: '0xd66d72c3e44037229d01cd28ec541c10d8989ff1',
    tickLens: '0xac80bb20b9f763de59557bb0d88936800da7620e',
    proxyAdmin: '0x9cd6a1f83e35f40910ae5b59d86e3eb03ca32d21',
    quoterV2: '0x90a15ed68e1bd5eb02d972de6dde6fa72b0ff3c7',
    nonfungibleTokenPositionDescriptorImpl: '0x318cedc9cd4ead794eb3ff09de67463a0a476425',
    nonfungibleTokenPositionDescriptor: '0x2ffa05a4e3a06b1cb3716e294a31bd10356cd2ac',
    nonfungiblePositionManager: '0xc421820a10be395d41a25873f45e27b5cea80dfe',
    swapRouter02: '0xee6188037ab4f10a071cfa549fde21dc415e78a3',
    uniswapV3Staker: '0xbfc8692eded1c443569c17e7e6618033d533931c',
    permit2: '0xfd1f634f6c4b8f7b63ed9b32e748a049b28e743f',
    unsupportedProtocol: '0xc09eec68cd04842b213507ad0e0acbfcce87316a',
    universalRouter: '0x6ab3a902b67f0caf98c4714557845840e9645a5c',
    pools: ['0x2135ea24a418c8312b302dfce8eb3897b4c68804'],
  }

  test('getContracts: should match snapshot', async () => {
    const verifier = new TestUniswapV3ContractVerifier(
      await createPublicClient({
        chain: holesky,
        transport: http('https://1rpc.io/holesky'),
      }),
      {
        apiKey: process.env.HOLESKY_ETHERSCAN_API_KEY ?? '',
        apiUrl: 'https://api-holesky.etherscan.io/api',
        url: 'https://holesky.etherscan.io',
      },
    )
    expect(verifier.testGetContracts(registry)).resolves.toMatchSnapshot()
  })
})
