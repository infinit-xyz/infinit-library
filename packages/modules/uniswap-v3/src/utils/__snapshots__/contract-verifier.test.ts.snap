// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UniswapV3ContractVerifier.ts > export: should match snapshot 1`] = `
[
  {
    "address": "0x8f0c29795615148f30bd402876fa69cbf88a91d4",
  },
  {
    "address": "0xd66d72c3e44037229d01cd28ec541c10d8989ff1",
  },
  {
    "address": "0xac80bb20b9f763de59557bb0d88936800da7620e",
  },
  {
    "address": "0x9cd6a1f83e35f40910ae5b59d86e3eb03ca32d21",
  },
  {
    "address": "0x90a15ed68e1bd5eb02d972de6dde6fa72b0ff3c7",
  },
  {
    "address": "0x318cedc9cd4ead794eb3ff09de67463a0a476425",
  },
  {
    "address": "0x2ffa05a4e3a06b1cb3716e294a31bd10356cd2ac",
  },
  {
    "address": "0xc421820a10be395d41a25873f45e27b5cea80dfe",
  },
  {
    "address": "0xee6188037ab4f10a071cfa549fde21dc415e78a3",
  },
  {
    "address": "0xbfc8692eded1c443569c17e7e6618033d533931c",
  },
  {
    "address": "0xfd1f634f6c4b8f7b63ed9b32e748a049b28e743f",
  },
  {
    "address": "0xc09eec68cd04842b213507ad0e0acbfcce87316a",
  },
  {
    "address": "0x6ab3a902b67f0caf98c4714557845840e9645a5c",
  },
  {
    "address": "0x2135ea24a418c8312b302dfce8eb3897b4c68804",
    "constructorArgs": [],
  },
]
`;
